src/quic/tokens.cc

```diff
@@ -0,0 +1,255 @@
+                                         version,
+                                         address.data(),
+                                         address.length(),
+                                         retry_cid,
+                                         odcid,
+                                         uv_hrtime());
+  DCHECK_GE(ret, 0);
+  DCHECK_LE(ret, RetryToken::kRetryTokenLen);
+  DCHECK_EQ(buffer[0], RetryToken::kTokenMagic);
+  // This shouldn't be possible but we handle it anyway just to be safe.
+  if (ret == 0) return {nullptr, 0};
+  return {buffer, static_cast<size_t>(ret)};
+}
+
+ngtcp2_vec generate_regular_token(uint8_t* buffer,
+                                  uint64_t version,
+                                  const SocketAddress& address,
+                                  const TokenSecret& token_secret) {
+  ssize_t ret =
+      ngtcp2_crypto_generate_regular_token(buffer,
+                                           token_secret,
+                                           TokenSecret::QUIC_TOKENSECRET_LEN,
+                                           address.data(),
+                                           address.length(),
+                                           uv_hrtime());
+  DCHECK_GE(ret, 0);
+  DCHECK_LE(ret, RegularToken::kRegularTokenLen);
+  DCHECK_EQ(buffer[0], RegularToken::kTokenMagic);
+  // This shouldn't be possible but we handle it anyway just to be safe.
+  if (ret == 0) return {nullptr, 0};
+  return {buffer, static_cast<size_t>(ret)};
+}
+}  // namespace
+
+RetryToken::RetryToken(uint64_t version,
+                       const SocketAddress& address,
+                       const CID& retry_cid,
+                       const CID& odcid,
+                       const TokenSecret& token_secret)
+    : buf_(),
+      ptr_(generate_retry_token(
+          buf_, version, address, retry_cid, odcid, token_secret)) {}
+
+RetryToken::RetryToken(const uint8_t* token, size_t size)
+    : ptr_(ngtcp2_vec{const_cast<uint8_t*>(token), size}) {
+  DCHECK_LE(size, RetryToken::kRetryTokenLen);
+  DCHECK_IMPLIES(token == nullptr, size = 0);
+}
+
+std::optional<CID> RetryToken::Validate(uint64_t version,
+                                        const SocketAddress& addr,
+                                        const CID& dcid,
+                                        const TokenSecret& token_secret,
+                                        uint64_t verification_expiration) {
+  if (ptr_.base == nullptr || ptr_.len == 0) return std::nullopt;
+  ngtcp2_cid ocid;
+  int ret = ngtcp2_crypto_verify_retry_token(
+      &ocid,
+      ptr_.base,
+      ptr_.len,
+      token_secret,
+      TokenSecret::QUIC_TOKENSECRET_LEN,
+      version,
+      addr.data(),
+      addr.length(),
+      dcid,
+      std::min(verification_expiration, QUIC_MIN_RETRYTOKEN_EXPIRATION),
+      uv_hrtime());
+  if (ret != 0) return std::nullopt;
+  return std::make_optional<CID>(ocid);
```

Comment:
No need to run `std::make_optional`. from my small research I learned that: `std::make_optional` only makes sense when you want to create `std::optional<std::optional<CID>>`.

```suggestion
  return std::optional<CID>(ocid);
```